### PR TITLE
Bug 1523379 - TPStatsBlocklists is not covered by tests

### PR DESCRIPTION
**Pull request by RobertPieta**
https://bugzilla.mozilla.org/show_bug.cgi?id=1523379

`TPStatsBlocklistsTests.swift` is added to provide benchmarks and a unit test for `TPStatsBlocklists`. Any recommendations or missed edge cases that should be added to the `urlIsInList` test can help a lot.

The motivation for covering `TPStatsBlocklists` with tests, in particular, is a followup performance improvement that benefits from having test coverage to ensure `TPStatsBlocklists` retains desired functionality.

### CODE REVIEW

File: ClientTests/TPStatsBlocklistsTests.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    
+    func testURLInListPerformance() {
+        blocklists.load()
+        
+        let whitelistedRegexs = [".google.com"].compactMap { (domain) -> NSRegularExpression? in
+            return wildcardContentBlockerDomainToRegex(domain: domain)
+        }
+        
+        self.measureMetrics([.wallClockTime], automaticallyStartMeasuring: true) {
+            for _ in 0..<100 {
```
Comment (garvankeeley) [inline ClientTests/TPStatsBlocklistsTests.swift]: This will run 1000x, how long does it take? Preferred to keep perf measurements in the 100ms range to keep test suites running quickly.

File: ClientTests/TPStatsBlocklistsTests.swift [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        XCTAssertEqual(blocklist("https://sub.aolanswers.com"), .content)
+        XCTAssertEqual(blocklist("https://aolanswers.com/track"), .content)
+        XCTAssertEqual(blocklist("https://aol.com.aolanswers.com"), .content)
+        XCTAssertEqual(blocklist("https://aol.com.aolanswers.com", ["ers.com"]), nil)
+        XCTAssertEqual(blocklist("https://games.com.aolanswers.com"), .content)
+        XCTAssertEqual(blocklist("https://bluesky.com.aolanswers.com"), .content)
+
+        XCTAssertEqual(blocklist("https://sub.xiti.com/track"), .analytics)
+        XCTAssertEqual(blocklist("https://backtype.com"), .social)
+        XCTAssertEqual(blocklist("https://backtype.com", [".firefox.com", "e.com"]), nil)
```
Comment (garvankeeley) [inline ClientTests/TPStatsBlocklistsTests.swift]: Nice! This reveals whitelisting is broken due to the conversion of the domain fragment to a regex. ("e.com" should only whitelist the "e" host.)

Comment (RobertPieta) [reply ClientTests/TPStatsBlocklistsTests.swift]: In the json files containing the rules the whitelist exceptions are stored as `*e.com` which `wildcardContentBlockerDomainToRegex` will turn into `\\.e\\.com`. I think that's what you want. I can update the test to use `*e.com` to better match the actual files.

File: ClientTests/TPStatsBlocklistsTests.swift (hunk 1 above)
Comment (RobertPieta) [reply ClientTests/TPStatsBlocklistsTests.swift]: Avg 0.003, though I think the cumulative start-to-finish test case (including setup, load, test,  teardown) takes 1.2s on my machine. Setting the loop to `0..<1` runs start-to-finish 1.15s but the avg is 0.000.

**Comment by garvankeeley** [comment]
The follow up bug https://bugzilla.mozilla.org/show_bug.cgi?id=1523389 describes how the domain whitelisting is broken, as revealed by these tests.

File: ClientTests/TPStatsBlocklistsTests.swift (hunk 2 above)
Comment (garvankeeley) [reply ClientTests/TPStatsBlocklistsTests.swift]: If you can update it that will be better as it won't look like a there is a bug in the code.

I am going to keep that follow-up bug filed, and explain more of my thinking, which is the code should be more defensive to handle unexpected domain inputs and not create wildcard regexs that could match a domain fragment.

File: ClientTests/TPStatsBlocklistsTests.swift (hunk 1 above)
Comment (garvankeeley) [reply ClientTests/TPStatsBlocklistsTests.swift]: A <50ms test SGTM, thanks for clarifyling!

File: ClientTests/TPStatsBlocklistsTests.swift (hunk 2 above)
Comment (garvankeeley) [reply ClientTests/TPStatsBlocklistsTests.swift]: I added more info to the followup bug https://bugzilla.mozilla.org/show_bug.cgi?id=1523389#c1

Comment (RobertPieta) [reply ClientTests/TPStatsBlocklistsTests.swift]: Updated!

**Comment by garvankeeley** [comment]
Can I close this, as these changes will land in #4501 ?

**Comment by RobertPieta** [comment]
> Can I close this, as these changes will land in #4501 ?

Sounds good to me!
